File: src/lib/config/module_spec.cc
// Copyright (C) 2010-2015 Internet Systems Consortium.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

#include <config/module_spec.h>

#include <sstream>
#include <iostream>
#include <fstream>
#include <cerrno>

#include <boost/foreach.hpp>

// todo: add more context to thrown ModuleSpecErrors?

using namespace isc::data;
using namespace isc::config;

namespace {
//
// Private functions
//

void
check_leaf_item(ConstElementPtr spec, const std::string& name,
                Element::types type, bool mandatory)
{
    if (spec->contains(name)) {
        if (type == Element::any || spec->get(name)->getType() == type) {
            return;
        } else {
            isc_throw(ModuleSpecError,
                      name + " not of type " + Element::typeToName(type));
        }
    } else if (mandatory) {
        // todo: want parent item name, and perhaps some info about location
        // in list? or just catch and throw new...
        // or make this part non-throwing and check return value...
        isc_throw(ModuleSpecError, name + " missing in " + spec->str());
    }
}

void check_config_item_list(ConstElementPtr spec);

void
check_config_item(ConstElementPtr spec) {
    check_leaf_item(spec, "item_name", Element::string, true);
    check_leaf_item(spec, "item_type", Element::string, true);
    check_leaf_item(spec, "item_optional", Element::boolean, true);
    check_leaf_item(spec, "item_default",
                    Element::nameToType(spec->get("item_type")->stringValue()),
                    !spec->get("item_optional")->boolValue()
                   );

    // if list, check the list specification
    if (Element::nameToType(spec->get("item_type")->stringValue()) == Element::list) {
        check_leaf_item(spec, "list_item_spec", Element::map, true);
        check_config_item(spec->get("list_item_spec"));
    }

    if (spec->get("item_type")->stringValue() == "map") {
        check_leaf_item(spec, "map_item_spec", Element::list, true);
        check_config_item_list(spec->get("map_item_spec"));
    } else if (spec->get("item_type")->stringValue() == "named_set") {
        check_leaf_item(spec, "named_set_item_spec", Element::map, true);
        check_config_item(spec->get("named_set_item_spec"));
    }
}

void
check_config_item_list(ConstElementPtr spec) {
    if (spec->getType() != Element::list) {
        isc_throw(ModuleSpecError, "config_data is not a list of elements");
    }
    BOOST_FOREACH(ConstElementPtr item, spec->listValue()) {
        check_config_item(item);
    }
}

// checks whether the given element is a valid statistics specification
// returns false if the specification is bad
bool
check_format(ConstElementPtr value, ConstElementPtr format_name) {
    typedef std::map<std::string, std::string> format_types;
    format_types time_formats;
    // TODO: should be added other format types if necessary
    time_formats.insert(
        format_types::value_type("date-time", "%Y-%m-%dT%H:%M:%SZ") );
    time_formats.insert(
        format_types::value_type("date", "%Y-%m-%d") );
    time_formats.insert(
        format_types::value_type("time", "%H:%M:%S") );
    BOOST_FOREACH (const format_types::value_type& f, time_formats) {
        if (format_name->stringValue() == f.first) {
            struct tm tm;
            std::vector<char> buf(32);
            memset(&tm, 0, sizeof(tm));
            // reverse check
            return (strptime(value->stringValue().c_str(),
                             f.second.c_str(), &tm) != NULL
                    && strftime(&buf[0], buf.size(),
                                f.second.c_str(), &tm) != 0
                    && strncmp(value->stringValue().c_str(),
                               &buf[0], buf.size()) == 0);
        }
    }
    return (false);
}

void check_statistics_item_list(ConstElementPtr spec);

void
check_statistics_item_list(ConstElementPtr spec) {
    if (spec->getType() != Element::list) {
        isc_throw(ModuleSpecError, "statistics is not a list of elements");
    }
    BOOST_FOREACH(ConstElementPtr item, spec->listValue()) {
        check_config_item(item);
        // additional checks for statistics
        check_leaf_item(item, "item_title", Element::string, true);
        check_leaf_item(item, "item_description", Element::string, true);
        check_leaf_item(item, "item_format", Element::string, false);
        // checks name of item_format and validation of item_default
        if (item->contains("item_format")
            && item->contains("item_default")) {
            if(!check_format(item->get("item_default"),
                             item->get("item_format"))) {
                isc_throw(ModuleSpecError,
                    "item_default not valid type of item_format");
            }
        }
    }
}

void
check_command(ConstElementPtr spec) {
    check_leaf_item(spec, "command_name", Element::string, true);
    check_leaf_item(spec, "command_args", Element::list, true);
    check_config_item_list(spec->get("command_args"));
}

void
check_command_list(ConstElementPtr spec) {
    if (spec->getType() != Element::list) {
        isc_throw(ModuleSpecError, "commands is not a list of elements");
    }
    BOOST_FOREACH(ConstElementPtr item, spec->listValue()) {
        check_command(item);
    }
}

void
check_data_specification(ConstElementPtr spec) {
    check_leaf_item(spec, "module_name", Element::string, true);
    check_leaf_item(spec, "module_description", Element::string, false);
    // config_data is not mandatory; module could just define
    // commands and have no config
    if (spec->contains("config_data")) {
        check_config_item_list(spec->get("config_data"));
    }
    if (spec->contains("commands")) {
        check_command_list(spec->get("commands"));
    }
    if (spec->contains("statistics")) {
        check_statistics_item_list(spec->get("statistics"));
    }
}

// checks whether the given element is a valid module specification
// throws a ModuleSpecError if the specification is bad
void
check_module_specification(ConstElementPtr def) {
    try {
        check_data_specification(def);
    } catch (const TypeError& te) {
        isc_throw(ModuleSpecError, te.what());
    }
}
}

namespace isc {
namespace config {
//
// Public functions
//

ModuleSpec::ModuleSpec(ConstElementPtr module_spec_element,
                       const bool check)
                       throw(ModuleSpecError)
                       
{
    module_specification = module_spec_element;
    if (check) {
        check_module_specification(module_specification);
    }
}

ConstElementPtr
ModuleSpec::getCommandsSpec() const {
    if (module_specification->contains("commands")) {
        return (module_specification->get("commands"));
    } else {
        return (ElementPtr());
    }
}

ConstElementPtr
ModuleSpec::getConfigSpec() const {
    if (module_specification->contains("config_data")) {
        return (module_specification->get("config_data"));
    } else {
        return (ElementPtr());
    }
}

ConstElementPtr
ModuleSpec::getStatisticsSpec() const {
    if (module_specification->contains("statistics")) {
        return (module_specification->get("statistics"));
    } else {
        return (ElementPtr());
    }
}

const std::string
ModuleSpec::getModuleName() const {
    return (module_specification->get("module_name")->stringValue());
}

const std::string
ModuleSpec::getModuleDescription() const {
    if (module_specification->contains("module_description")) {
        return (module_specification->get("module_description")->stringValue());
    } else {
        return (std::string(""));
    }
}

bool
ModuleSpec::validateConfig(ConstElementPtr data, const bool full) const {
    ConstElementPtr spec = module_specification->find("config_data");
    return (validateSpecList(spec, data, full, ElementPtr()));
}

bool
ModuleSpec::validateStatistics(ConstElementPtr data, const bool full) const {
    ConstElementPtr spec = module_specification->find("statistics");
    return (validateSpecList(spec, data, full, ElementPtr()));
}

bool
ModuleSpec::validateCommand(const std::string& command,
                             ConstElementPtr args,
                             ElementPtr errors) const
{
    if (args->getType() != Element::map) {
        errors->add(Element::create("args for command " +
                                    command + " is not a map"));
        return (false);
    }

    ConstElementPtr commands_spec = module_specification->find("commands");
    if (!commands_spec) {
        // there are no commands according to the spec.
        errors->add(Element::create("The given module has no commands"));
        return (false);
    }

    BOOST_FOREACH(ConstElementPtr cur_command, commands_spec->listValue()) {
        if (cur_command->get("command_name")->stringValue() == command) {
            return (validateSpecList(cur_command->get("command_args"),
                                       args, true, errors));
        }
    }

    // this command is unknown
    errors->add(Element::create("Unknown command " + command));
    return (false);
}

bool
ModuleSpec::validateConfig(ConstElementPtr data, const bool full,
                            ElementPtr errors) const
{
    ConstElementPtr spec = module_specification->find("config_data");
    return (validateSpecList(spec, data, full, errors));
}

bool
ModuleSpec::validateStatistics(ConstElementPtr data, const bool full,
                               ElementPtr errors) const
{
    ConstElementPtr spec = module_specification->find("statistics");
    return (validateSpecList(spec, data, full, errors));
}

ModuleSpec
moduleSpecFromFile(const std::string& file_name, const bool check)
                   throw(JSONError, ModuleSpecError)
{
    std::ifstream file;

    // zero out the errno to be safe
    errno = 0;

    file.open(file_name.c_str());
    if (!file) {
        std::stringstream errs;
        errs << "Error opening " << file_name << ": " << strerror(errno);
        isc_throw(ModuleSpecError, errs.str());
    }

    ConstElementPtr module_spec_element = Element::fromJSON(file, file_name);
    if (module_spec_element->contains("module_spec")) {
        return (ModuleSpec(module_spec_element->get("module_spec"), check));
    } else {
        isc_throw(ModuleSpecError, "No module_spec in specification");
    }
}

ModuleSpec
moduleSpecFromFile(std::ifstream& in, const bool check)
                   throw(JSONError, ModuleSpecError)
{
    ConstElementPtr module_spec_element = Element::fromJSON(in);
    if (module_spec_element->contains("module_spec")) {
        return (ModuleSpec(module_spec_element->get("module_spec"), check));
    } else {
        isc_throw(ModuleSpecError, "No module_spec in specification");
    }
}


namespace {
//
// private functions
//

//
// helper functions for validation
//
bool
check_type(ConstElementPtr spec, ConstElementPtr element) {
    std::string cur_item_type;
    cur_item_type = spec->get("item_type")->stringValue();
    if (cur_item_type == "any") {
        return (true);
    }
    switch (element->getType()) {
        case Element::integer:
            return (cur_item_type == "integer");
            break;
        case Element::real:
            return (cur_item_type == "real");
            break;
        case Element::boolean:
            return (cur_item_type == "boolean");
            break;
        case Element::string:
            return (cur_item_type == "string");
            break;
        case Element::list:
            return (cur_item_type == "list");
            break;
        case Element::map:
            return (cur_item_type == "map" ||
                    cur_item_type == "named_set");
            break;
    }
    return (false);
}
}

bool
ModuleSpec::validateItem(ConstElementPtr spec, ConstElementPtr data,
                          const bool full, ElementPtr errors) const
{
    if (!check_type(spec, data)) {
        // we should do some proper error feedback here
        // std::cout << "type mismatch; not " << spec->get("item_type") << ": " << data << std::endl;
        // std::cout << spec << std::endl;
        if (errors) {
            errors->add(Element::create("Type mismatch"));
        }
        return (false);
    }
    if (data->getType() == Element::list) {
        ConstElementPtr list_spec = spec->get("list_item_spec");
        BOOST_FOREACH(ConstElementPtr list_el, data->listValue()) {
            if (!check_type(list_spec, list_el)) {
                if (errors) {
                    errors->add(Element::create("Type mismatch"));
                }
                return (false);
            }
            if (list_spec->get("item_type")->stringValue() == "map") {
                if (!validateItem(list_spec, list_el, full, errors)) {
                    return (false);
                }
            }
        }
    }
    if (data->getType() == Element::map) {
        // either a normal 'map' or a 'named set' (determined by which
        // subspecification it has)
        if (spec->contains("map_item_spec")) {
            if (!validateSpecList(spec->get("map_item_spec"), data, full, errors)) {
                return (false);
            }
        } else {
            typedef std::pair<std::string, ConstElementPtr> maptype;

            BOOST_FOREACH(maptype m, data->mapValue()) {
                if (!validateItem(spec->get("named_set_item_spec"), m.second, full, errors)) {
                    return (false);
                }
            }
        }
    }
    if (spec->contains("item_format")) {
        if (!check_format(data, spec->get("item_format"))) {
            if (errors) {
                errors->add(Element::create("Format mismatch"));
            }
            return (false);
        }
    }
    return (true);
}

// spec is a map with item_name etc, data is a map
bool
ModuleSpec::validateSpec(ConstElementPtr spec, ConstElementPtr data,
                          const bool full, ElementPtr errors) const
{
    std::string item_name = spec->get("item_name")->stringValue();
    bool optional = spec->get("item_optional")->boolValue();
    ConstElementPtr data_el;
    data_el = data->get(item_name);
    
    if (data_el) {
        if (!validateItem(spec, data_el, full, errors)) {
            return (false);
        }
    } else {
        if (!optional && full) {
            if (errors) {
                errors->add(Element::create("Non-optional value missing"));
            }
            return (false);
        }
    }
    return (true);
}

// spec is a list of maps, data is a map
bool
ModuleSpec::validateSpecList(ConstElementPtr spec, ConstElementPtr data,
                               const bool full, ElementPtr errors) const
{
    bool validated = true;
    BOOST_FOREACH(ConstElementPtr cur_spec_el, spec->listValue()) {
        if (!validateSpec(cur_spec_el, data, full, errors)) {
            validated = false;
        }
    }

    typedef std::pair<std::string, ConstElementPtr> maptype;
    
    BOOST_FOREACH(maptype m, data->mapValue()) {
        // Ignore 'version' as a config element
        if (m.first.compare("version") != 0) {
            bool found = false;
            BOOST_FOREACH(ConstElementPtr cur_spec_el, spec->listValue()) {
                if (cur_spec_el->get("item_name")->stringValue().compare(m.first) == 0) {
                    found = true;
                }
            }
            if (!found) {
                validated = false;
                if (errors) {
                    errors->add(Element::create("Unknown item " + m.first));
                }
            }
        }
    }

    return (validated);
}

}
}

File: src/lib/python/isc/config/module_spec.py
# Copyright (C) 2009  Internet Systems Consortium.
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SYSTEMS CONSORTIUM
# DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL
# INTERNET SYSTEMS CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
# FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
# NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
# WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

"""Module Specifications

   A module specification holds the information about what configuration
   a module can have, and what commands it understands. It provides
   functions to read it from a .spec file, and to validate a given
   set of data against the specification
"""

import json
import sys

import isc.cc.data

# file objects are passed around as _io.TextIOWrapper objects
# import that so we can check those types

class ModuleSpecError(Exception):
    """This exception is raised it the ModuleSpec fails to initialize
       or if there is a failure or parse error reading the specification
       file"""
    pass

def module_spec_from_file(spec_file, check = True):
    """Returns a ModuleSpec object defined by the file at spec_file.
       If check is True, the contents are verified. If there is an error
       in those contents, a ModuleSpecError is raised.
       A ModuleSpecError is also raised if the file cannot be read, or
       if it is not valid JSON."""
    module_spec = None
    try:
        if hasattr(spec_file, 'read'):
            json_str = spec_file.read()
            module_spec = json.loads(json_str)
        elif type(spec_file) == str:
            file = open(spec_file)
            json_str = file.read()
            module_spec = json.loads(json_str)
            file.close()
        else:
            raise ModuleSpecError("spec_file not a str or file-like object")
    except ValueError as ve:
        raise ModuleSpecError("JSON parse error: " + str(ve))
    except IOError as ioe:
        raise ModuleSpecError("JSON read error: " + str(ioe))

    if 'module_spec' not in module_spec:
        raise ModuleSpecError("Data definition has no module_spec element")

    result = ModuleSpec(module_spec['module_spec'], check)
    return result

class ModuleSpec:
    def __init__(self, module_spec, check = True):
        """Initializes a ModuleSpec object from the specification in
           the given module_spec (which must be a dict). If check is
           True, the contents are verified. Raises a ModuleSpec error
           if there is something wrong with the contents of the dict"""
        if type(module_spec) != dict:
            raise ModuleSpecError("module_spec is of type " + str(type(module_spec)) + ", not dict")
        if check:
            _check(module_spec)
        self._module_spec = module_spec

    def validate_config(self, full, data, errors = None):
        """Check whether the given piece of data conforms to this
           data definition. If so, it returns True. If not, it will
           return false. If errors is given, and is an array, a string
           describing the error will be appended to it. The current
           version stops as soon as there is one error so this list
           will not be exhaustive. If 'full' is true, it also errors on
           non-optional missing values. Set this to False if you want to
           validate only a part of a configuration tree (like a list of
           non-default values)"""
        data_def = self.get_config_spec()
        if data_def:
            return _validate_spec_list(data_def, full, data, errors)
        else:
            # no spec, always bad
            if errors != None:
                errors.append("No config_data specification")
            return False

    def validate_command(self, cmd_name, cmd_params, errors = None):
        '''Check whether the given piece of command conforms to this 
        command definition. If so, it reutrns True. If not, it will 
        return False. If errors is given, and is an array, a string
        describing the error will be appended to it. The current version
        stops as soon as there is one error.
           cmd_name is command name to be validated, cmd_params includes 
        command's parameters needs to be validated. cmd_params must 
        be a map, with the format like:
        {param1_name: param1_value, param2_name: param2_value}
        '''
        cmd_spec = self.get_commands_spec()
        if not cmd_spec:
            return False

        for cmd in cmd_spec:
            if cmd['command_name'] != cmd_name:
                continue
            return _validate_spec_list(cmd['command_args'], True, cmd_params, errors)

        return False

    def get_module_name(self):
        """Returns a string containing the name of the module as
           specified by the specification given at __init__()"""
        return self._module_spec['module_name']

    def get_module_description(self):
        """Returns a string containing the description of the module as
           specified by the specification given at __init__().
           Returns an empty string if there is no description.
        """
        if 'module_description' in self._module_spec:
            return self._module_spec['module_description']
        else:
            return ""

    def get_full_spec(self):
        """Returns a dict representation of the full module specification"""
        return self._module_spec

    def get_config_spec(self):
        """Returns a dict representation of the configuration data part
           of the specification, or None if there is none."""
        if 'config_data' in self._module_spec:
            return self._module_spec['config_data']
        else:
            return None
    
    def get_commands_spec(self):
        """Returns a dict representation of the commands part of the
           specification, or None if there is none."""
        if 'commands' in self._module_spec:
            return self._module_spec['commands']
        else:
            return None
    
    def __str__(self):
        """Returns a string representation of the full specification"""
        return self._module_spec.__str__()

def _check(module_spec):
    """Checks the full specification. This is a dict that contains the
       element "module_spec", which is in itself a dict that
       must contain at least a "module_name" (string) and optionally
       a "config_data" and a "commands" element, both of which are lists
       of dicts. Raises a ModuleSpecError if there is a problem."""
    if type(module_spec) != dict:
        raise ModuleSpecError("data specification not a dict")
    if "module_name" not in module_spec:
        raise ModuleSpecError("no module_name in module_spec")
    if "module_description" in module_spec and \
       type(module_spec["module_description"]) != str:
        raise ModuleSpecError("module_description is not a string")
    if "config_data" in module_spec:
        _check_config_spec(module_spec["config_data"])
    if "commands" in module_spec:
        _check_command_spec(module_spec["commands"])

def _check_config_spec(config_data):
    # config data is a list of items represented by dicts that contain
    # things like "item_name", depending on the type they can have
    # specific subitems
    """Checks a list that contains the configuration part of the
       specification. Raises a ModuleSpecError if there is a
       problem."""
    if type(config_data) != list:
        raise ModuleSpecError("config_data is of type " + str(type(config_data)) + ", not a list of items")
    for config_item in config_data:
        _check_item_spec(config_item)

def _check_command_spec(commands):
    """Checks the list that contains a set of commands. Raises a
       ModuleSpecError is there is an error"""
    if type(commands) != list:
        raise ModuleSpecError("commands is not a list of commands")
    for command in commands:
        if type(command) != dict:
            raise ModuleSpecError("command in commands list is not a dict")
        if "command_name" not in command:
            raise ModuleSpecError("no command_name in command item")
        command_name = command["command_name"]
        if type(command_name) != str:
            raise ModuleSpecError("command_name not a string: " + str(type(command_name)))
        if "command_description" in command:
            if type(command["command_description"]) != str:
                raise ModuleSpecError("command_description not a string in " + command_name)
        if "command_args" in command:
            if type(command["command_args"]) != list:
                raise ModuleSpecError("command_args is not a list in " + command_name)
            for command_arg in command["command_args"]:
                if type(command_arg) != dict:
                    raise ModuleSpecError("command argument not a dict in " + command_name)
                _check_item_spec(command_arg)
        else:
            raise ModuleSpecError("command_args missing in " + command_name)
    pass

def _check_item_spec(config_item):
    """Checks the dict that defines one config item
       (i.e. containing "item_name", "item_type", etc.
       Raises a ModuleSpecError if there is an error"""
    if type(config_item) != dict:
        raise ModuleSpecError("item spec not a dict")
    if "item_name" not in config_item:
        raise ModuleSpecError("no item_name in config item")
    if type(config_item["item_name"]) != str:
        raise ModuleSpecError("item_name is not a string: " + str(config_item["item_name"]))
    item_name = config_item["item_name"]
    if "item_type" not in config_item:
        raise ModuleSpecError("no item_type in config item")
    item_type = config_item["item_type"]
    if type(item_type) != str:
        raise ModuleSpecError("item_type in " + item_name + " is not a string: " + str(type(item_type)))
    if item_type not in ["integer", "real", "boolean", "string", "list", "map", "any"]:
        raise ModuleSpecError("unknown item_type in " + item_name + ": " + item_type)
    if "item_optional" in config_item:
        if type(config_item["item_optional"]) != bool:
            raise ModuleSpecError("item_default in " + item_name + " is not a boolean")
        if not config_item["item_optional"] and "item_default" not in config_item:
            raise ModuleSpecError("no default value for non-optional item " + item_name)
    else:
        raise ModuleSpecError("item_optional not in item " + item_name)
    if "item_default" in config_item:
        item_default = config_item["item_default"]
        if (item_type == "integer" and type(item_default) != int) or \
           (item_type == "real" and type(item_default) != float) or \
           (item_type == "boolean" and type(item_default) != bool) or \
           (item_type == "string" and type(item_default) != str) or \
           (item_type == "list" and type(item_default) != list) or \
           (item_type == "map" and type(item_default) != dict):
            raise ModuleSpecError("Wrong type for item_default in " + item_name)
    # TODO: once we have check_type, run the item default through that with the list|map_item_spec
    if item_type == "list":
        if "list_item_spec" not in config_item:
            raise ModuleSpecError("no list_item_spec in list item " + item_name)
        if type(config_item["list_item_spec"]) != dict:
            raise ModuleSpecError("list_item_spec in " + item_name + " is not a dict")
        _check_item_spec(config_item["list_item_spec"])
    if item_type == "map":
        if "map_item_spec" not in config_item:
            raise ModuleSpecError("no map_item_sepc in map item " + item_name)
        if type(config_item["map_item_spec"]) != list:
            raise ModuleSpecError("map_item_spec in " + item_name + " is not a list")
        for map_item in config_item["map_item_spec"]:
            if type(map_item) != dict:
                raise ModuleSpecError("map_item_spec element is not a dict")
            _check_item_spec(map_item)


def _validate_type(spec, value, errors):
    """Returns true if the value is of the correct type given the
       specification"""
    data_type = spec['item_type']
    if data_type == "integer" and type(value) != int:
        if errors != None:
            errors.append(str(value) + " should be an integer")
        return False
    elif data_type == "real" and type(value) != float:
        if errors != None:
            errors.append(str(value) + " should be a real")
        return False
    elif data_type == "boolean" and type(value) != bool:
        if errors != None:
            errors.append(str(value) + " should be a boolean")
        return False
    elif data_type == "string" and type(value) != str:
        if errors != None:
            errors.append(str(value) + " should be a string")
        return False
    elif data_type == "list" and type(value) != list:
        if errors != None:
            errors.append(str(value) + " should be a list")
        return False
    elif data_type == "map" and type(value) != dict:
        if errors != None:
            errors.append(str(value) + " should be a map")
        return False
    else:
        return True

def _validate_item(spec, full, data, errors):
    if not _validate_type(spec, data, errors):
        return False
    elif type(data) == list:
        list_spec = spec['list_item_spec']
        for data_el in data:
            if not _validate_type(list_spec, data_el, errors):
                return False
            if list_spec['item_type'] == "map":
                if not _validate_item(list_spec, full, data_el, errors):
                    return False
    elif type(data) == dict:
        if not _validate_spec_list(spec['map_item_spec'], full, data, errors):
            return False
    return True

def _validate_spec(spec, full, data, errors):
    item_name = spec['item_name']
    item_optional = spec['item_optional']

    if item_name in data:
        return _validate_item(spec, full, data[item_name], errors)
    elif full and not item_optional:
        if errors != None:
            errors.append("non-optional item " + item_name + " missing")
        return False
    else:
        return True

def _validate_spec_list(module_spec, full, data, errors):
    for spec_item in module_spec:
        if not _validate_spec(spec_item, full, data, errors):
            return False
    return True
